File: frontend/src/lib/rules.ts
import type { Rule } from "./api";

// Mirrors backend app/rules.py — returns human-readable violations ([] if compliant).
export function checkRules(rules: Rule[], choices: string[]): string[] {
  const errors: string[] = [];
  for (const rule of rules ?? []) {
    if (rule.type === "position_in") {
      const subs = rule.subjects;
      if (rule.position >= 1 && rule.position <= choices.length) {
        if (!subs.includes(choices[rule.position - 1])) {
          errors.push(`Choice ${rule.position} must be one of: ${subs.join(", ")}`);
        }
      }
    } else if (rule.type === "require_one_of") {
      const have = choices.filter((c) => rule.subjects.includes(c)).length;
      if (have < rule.min) {
        errors.push(`Pick at least ${rule.min} of: ${rule.subjects.join(", ")}`);
      }
    } else if (rule.type === "only_at") {
      choices.forEach((c, idx) => {
        if (rule.subjects.includes(c) && !rule.positions.includes(idx + 1)) {
          errors.push(`${c} may only be chosen at position(s): ${rule.positions.join(", ")}`);
        }
      });
    }
  }
  return errors;
}

// Subjects selectable at a given 1-based choice position, given the rules.
export function allowedAtPosition(
  rules: Rule[],
  position: number,
  all: string[],
): string[] {
  let allowed = all;
  for (const rule of rules ?? []) {
    if (rule.type === "position_in" && rule.position === position) {
      allowed = allowed.filter((s) => rule.subjects.includes(s));
    }
    if (rule.type === "only_at" && !rule.positions.includes(position)) {
      // Subjects locked to other positions can't appear here.
      allowed = allowed.filter((s) => !rule.subjects.includes(s));
    }
  }
  return allowed;
}

File: backend/app/rules.py
"""Choice rules (item 14). Rules are stored on the timetable as a list of dicts and
enforced as HARD limits wherever choices are entered (student submit, teacher add /
edit / CSV import) — so only rule-compliant choices ever reach the solver.

Rule types (each a dict in TimetableModel.rules):
  {"type": "position_in",   "position": 1, "subjects": [...]}  choice N must be in set
  {"type": "require_one_of", "subjects": [...], "min": 1}       at least N of set chosen
  {"type": "only_at",        "subjects": [...], "positions": [1]} subjects allowed only
                                                                 at these choice positions
Rules apply to the ranked CHOICES (by position), not backups.
"""
from __future__ import annotations


def check_rules(rules: list[dict], choices: list[str]) -> list[str]:
    """Return a list of human-readable violations ([] if all satisfied)."""
    errors: list[str] = []
    for rule in rules or []:
        t = rule.get("type")
        subs = set(rule.get("subjects") or [])
        if t == "position_in":
            pos = int(rule.get("position", 0))
            if 1 <= pos <= len(choices) and choices[pos - 1] not in subs:
                errors.append(
                    f"Choice {pos} must be one of: {', '.join(sorted(subs))}"
                )
        elif t == "require_one_of":
            need = int(rule.get("min", 1))
            have = sum(1 for c in choices if c in subs)
            if have < need:
                errors.append(
                    f"Pick at least {need} of: {', '.join(sorted(subs))}"
                )
        elif t == "only_at":
            positions = {int(p) for p in rule.get("positions") or []}
            for idx, c in enumerate(choices, start=1):
                if c in subs and idx not in positions:
                    allowed = ", ".join(str(p) for p in sorted(positions))
                    errors.append(f"{c} may only be chosen at position(s): {allowed}")
    return errors


def rules_error(rules: list[dict], choices: list[str]) -> str | None:
    """Single combined message, or None when compliant."""
    errors = check_rules(rules, choices)
    return "; ".join(errors) if errors else None
